File: goblinframework-core/src/main/java/org/goblinframework/core/util/ClassUtils.java
package org.goblinframework.core.util;

import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * Class related miscellaneous utilities.
 */
abstract public class ClassUtils extends org.apache.commons.lang3.ClassUtils {

  @NotNull
  public static ClassLoader getDefaultClassLoader() {
    ClassLoader classLoader = org.springframework.util.ClassUtils.getDefaultClassLoader();
    if (classLoader == null) {
      throw new UnsupportedOperationException("Cannot access ClassLoader");
    }
    return classLoader;
  }

  @NotNull
  public static Class<?> filterCglibProxyClass(@NotNull Class<?> clazz) {
    Class<?> clazzForUse = clazz;
    while (org.springframework.util.ClassUtils.isCglibProxyClass(clazzForUse)) {
      clazzForUse = clazzForUse.getSuperclass();
    }
    return clazzForUse;
  }

  @NotNull
  public static Class<?> loadClass(@NotNull String className) throws ClassNotFoundException {
    return getClass(getDefaultClassLoader(), className);
  }

  public static List<Class<?>> getClassInheritanceHierarchy(@NotNull final Class<?> clazz, boolean topEndian) {
    Class<?> theClass = clazz;
    List<Class<?>> chain = new LinkedList<>();
    chain.add(theClass);
    while ((theClass = theClass.getSuperclass()) != null) {
      chain.add(theClass);
    }
    if (topEndian) {
      Collections.reverse(chain);
    }
    return Collections.unmodifiableList(chain);
  }
}

File: goblin-core/src/main/java/org/goblinframework/core/util/_ClassUtils.kt
package org.goblinframework.core.util

/**
 * Return the default ClassLoader, typically try to use the thread context ClassLoader first.
 */
fun getDefaultClassLoader(): ClassLoader {
  val classLoader = org.springframework.util.ClassUtils.getDefaultClassLoader()
  requireNotNull(classLoader)
  return classLoader
}

/**
 * Load the class represented by specified className using the default ClassLoader.
 */
@Throws(ClassNotFoundException::class)
fun loadClass(className: String, initialize: Boolean = true): Class<*> {
  return org.apache.commons.lang3.ClassUtils.getClass(getDefaultClassLoader(), className, initialize)
}

